File: sorting/insertionsort.php
<?php
// Insertion Sort Algorithm Implementation in PHP
//
// Time Complexity:
// - Best Case: O(n) - when array is already sorted
// - Average Case: O(n²)
// - Worst Case: O(n²) - when array is reverse sorted
// Space Complexity: O(1) for in-place, O(n) for functional approach
//
// Insertion Sort builds the final sorted array one item at a time.
//
// PHP features:
// - Type declarations (PHP 7+)
// - Static methods in classes
// - Array functions and functional programming
// - Callable comparators
// - Namespace organization

namespace AlgorithmsMultiverse\Sorting;

/**
 * Insertion Sort implementation with multiple variants
 */
class InsertionSort
{
    // MARK: - Basic Insertion Sort

    /**
     * Performs standard insertion sort.
     *
     * Time Complexity: O(n²) average and worst case, O(n) best case
     * Space Complexity: O(n) for the new array
     *
     * Example visualization:
     *   Initial: [5, 2, 8, 6, 1]
     *   Step 1:  [2, 5, 8, 6, 1]  // Insert 2
     *   Step 2:  [2, 5, 8, 6, 1]  // 8 already in place
     *   Step 3:  [2, 5, 6, 8, 1]  // Insert 6
     *   Step 4:  [1, 2, 5, 6, 8]  // Insert 1
     *
     * @param array $arr The array to sort
     * @return array The sorted array
     */
    public static function sort(array $arr): array
    {
        if (count($arr) <= 1) {
            return $arr;
        }

        $result = $arr;
        self::sortInPlace($result);
        return $result;
    }

    /**
     * Sorts the array in-place using insertion sort.
     *
     * Time Complexity: O(n²) average and worst case, O(n) best case
     * Space Complexity: O(1)
     *
     * @param array &$arr The array to sort (passed by reference)
     */
    public static function sortInPlace(array &$arr): void
    {
        $n = count($arr);

        for ($i = 1; $i < $n; $i++) {
            $key = $arr[$i];
            $j = $i - 1;

            // Move elements greater than key one position ahead
            while ($j >= 0 && $arr[$j] > $key) {
                $arr[$j + 1] = $arr[$j];
                $j--;
            }

            $arr[$j + 1] = $key;
        }
    }

    /**
     * Insertion sort with custom comparator.
     *
     * @param array $arr The array to sort
     * @param callable $comparator Comparison function
     * @return array The sorted array
     */
    public static function sortWithComparator(array $arr, callable $comparator): array
    {
        if (count($arr) <= 1) {
            return $arr;
        }

        $result = $arr;
        $n = count($result);

        for ($i = 1; $i < $n; $i++) {
            $key = $result[$i];
            $j = $i - 1;

            while ($j >= 0 && $comparator($key, $result[$j]) < 0) {
                $result[$j + 1] = $result[$j];
                $j--;
            }

            $result[$j + 1] = $key;
        }

        return $result;
    }

    // MARK: - Recursive Insertion Sort

    /**
     * Performs recursive insertion sort.
     *
     * Time Complexity: O(n²)
     * Space Complexity: O(n) for recursion stack
     *
     * @param array $arr The array to sort
     * @return array The sorted array
     */
    public static function sortRecursive(array $arr): array
    {
        if (count($arr) <= 1) {
            return $arr;
        }

        $result = $arr;
        self::sortRecursiveHelper($result, count($result));
        return $result;
    }

    /**
     * Helper function for recursive insertion sort.
     *
     * @param array &$arr The array to sort
     * @param int $n The number of elements to sort
     */
    private static function sortRecursiveHelper(array &$arr, int $n): void
    {
        // Base case
        if ($n <= 1) {
            return;
        }

        // Sort first n-1 elements
        self::sortRecursiveHelper($arr, $n - 1);

        // Insert last element at its correct position
        $key = $arr[$n - 1];
        $j = $n - 2;

        while ($j >= 0 && $arr[$j] > $key) {
            $arr[$j + 1] = $arr[$j];
            $j--;
        }

        $arr[$j + 1] = $key;
    }

    // MARK: - Binary Insertion Sort

    /**
     * Uses binary search to find insertion position.
     *
     * Time Complexity: O(n²) for moves, O(n log n) for comparisons
     * Space Complexity: O(n)
     *
     * @param array $arr The array to sort
     * @return array The sorted array
     */
    public static function binaryInsertionSort(array $arr): array
    {
        if (count($arr) <= 1) {
            return $arr;
        }

        $result = $arr;
        $n = count($result);

        for ($i = 1; $i < $n; $i++) {
            $key = $result[$i];

            // Find position using binary search
            $pos = self::binarySearchPosition($result, 0, $i - 1, $key);

            // Shift elements to make space
            for ($j = $i - 1; $j >= $pos; $j--) {
                $result[$j + 1] = $result[$j];
            }

            $result[$pos] = $key;
        }

        return $result;
    }

    /**
     * Binary search to find the correct insertion position.
     *
     * @param array $arr The array
     * @param int $left Left boundary
     * @param int $right Right boundary
     * @param mixed $key The key to insert
     * @return int The position to insert
     */
    private static function binarySearchPosition(array $arr, int $left, int $right, $key): int
    {
        if ($right <= $left) {
            return ($key > $arr[$left]) ? $left + 1 : $left;
        }

        $mid = intdiv($left + $right, 2);

        if ($key == $arr[$mid]) {
            return $mid + 1;
        }

        if ($key > $arr[$mid]) {
            return self::binarySearchPosition($arr, $mid + 1, $right, $key);
        }

        return self::binarySearchPosition($arr, $left, $mid - 1, $key);
    }

    // MARK: - Shell Sort

    /**
     * Performs shell sort (generalization of insertion sort).
     *
     * Time Complexity: Depends on gap sequence (O(n log²n) for good sequences)
     * Space Complexity: O(n)
     *
     * @param array $arr The array to sort
     * @return array The sorted array
     */
    public static function shellSort(array $arr): array
    {
        if (count($arr) <= 1) {
            return $arr;
        }

        $result = $arr;
        $n = count($result);

        // Start with a large gap, then reduce (Knuth's sequence)
        $gap = 1;
        while ($gap < intdiv($n, 3)) {
            $gap = 3 * $gap + 1;
        }

        // Perform gapped insertion sort
        while ($gap > 0) {
            for ($i = $gap; $i < $n; $i++) {
                $key = $result[$i];
                $j = $i;

                // Insertion sort with gap
                while ($j >= $gap && $result[$j - $gap] > $key) {
                    $result[$j] = $result[$j - $gap];
                    $j -= $gap;
                }

                $result[$j] = $key;
            }

            $gap = intdiv($gap, 3);
        }

        return $result;
    }

    // MARK: - Sort Statistics

    /**
     * Insertion sort with statistics tracking
     *
     * @param array $arr The array to sort
     * @param SortStatistics $stats Statistics object to track operations
     * @return array The sorted array
     */
    public static function sortWithStats(array $arr, SortStatistics $stats): array
    {
        $stats->reset();
        if (count($arr) <= 1) {
            return $arr;
        }

        $result = $arr;
        $n = count($result);

        for ($i = 1; $i < $n; $i++) {
            $key = $result[$i];
            $j = $i - 1;

            while ($j >= 0) {
                $stats->comparisons++;
                if ($result[$j] > $key) {
                    $result[$j + 1] = $result[$j];
                    $stats->swaps++;
                    $j--;
                } else {
                    break;
                }
            }

            $result[$j + 1] = $key;
        }

        return $result;
    }

    // MARK: - Visualization

    /**
     * Creates a step-by-step visualization of insertion sort
     *
     * @param array $arr The array to visualize
     * @return array Array of strings showing each step
     */
    public static function visualizeInsertionSort(array $arr): array
    {
        $steps = [];
        $result = $arr;

        $steps[] = "Initial: [" . implode(", ", $result) . "]";

        $n = count($result);
        for ($i = 1; $i < $n; $i++) {
            $key = $result[$i];
            $j = $i - 1;

            $steps[] = "\nStep $i: Inserting $key";
            $steps[] = "  Before: [" . implode(", ", $result) . "]";

            while ($j >= 0 && $result[$j] > $key) {
                $result[$j + 1] = $result[$j];
                $j--;
            }

            $result[$j + 1] = $key;
            $steps[] = "  After:  [" . implode(", ", $result) . "]";
        }

        $steps[] = "\nFinal: [" . implode(", ", $result) . "]";
        return $steps;
    }

    // MARK: - Stability Demonstration

    /**
     * Demonstrates that insertion sort is stable
     */
    public static function demonstrateStability(): void
    {
        $data = [
            ['value' => 3, 'originalIndex' => 0],
            ['value' => 1, 'originalIndex' => 1],
            ['value' => 3, 'originalIndex' => 2],
            ['value' => 2, 'originalIndex' => 3],
            ['value' => 3, 'originalIndex' => 4],
        ];

        // Sort by value only
        $sorted = self::sortWithComparator($data, function($a, $b) {
            return $a['value'] <=> $b['value'];
        });

        echo "Stability Demonstration:\n";
        echo "Original: ";
        foreach ($data as $p) {
            echo "({$p['value']},{$p['originalIndex']}) ";
        }
        echo "\n";

        echo "Sorted:   ";
        foreach ($sorted as $p) {
            echo "({$p['value']},{$p['originalIndex']}) ";
        }
        echo "\n";

        // Check stability - all 3's should maintain original order
        $threeIndices = array_map(
            fn($p) => $p['originalIndex'],
            array_filter($sorted, fn($p) => $p['value'] === 3)
        );
        $isStable = $threeIndices == [0, 2, 4];
        echo "Stable: " . ($isStable ? "true" : "false") .
             " (indices of 3's: [" . implode(", ", $threeIndices) . "])\n";
    }

    // MARK: - Helper Functions

    /**
     * Check if an array is sorted in ascending order
     *
     * @param array $arr The array to check
     * @return bool True if sorted, false otherwise
     */
    public static function isSorted(array $arr): bool
    {
        $n = count($arr);
        for ($i = 0; $i < $n - 1; $i++) {
            if ($arr[$i] > $arr[$i + 1]) {
                return false;
            }
        }
        return true;
    }
}

/**
 * Tracks sorting operations
 */
class SortStatistics
{
    public int $comparisons = 0;
    public int $swaps = 0;

    public function reset(): void
    {
        $this->comparisons = 0;
        $this->swaps = 0;
    }

    public function __toString(): string
    {
        return "Comparisons: {$this->comparisons}, Swaps: {$this->swaps}";
    }
}

// MARK: - Demonstration and Testing

function demonstrateInsertionSort(): void
{
    echo "📝 Insertion Sort Implementation in PHP\n";
    echo str_repeat("=", 60) . "\n";

    // Test data
    $testCases = [
        ['arr' => [64, 34, 25, 12, 22, 11, 90], 'desc' => 'Random array'],
        ['arr' => [5, 2, 8, 6, 1, 9, 4], 'desc' => 'Small random array'],
        ['arr' => [1], 'desc' => 'Single element'],
        ['arr' => [], 'desc' => 'Empty array'],
        ['arr' => [3, 3, 3, 3, 3], 'desc' => 'All duplicates'],
        ['arr' => [9, 8, 7, 6, 5, 4, 3, 2, 1], 'desc' => 'Reverse sorted'],
        ['arr' => [1, 2, 3, 4, 5], 'desc' => 'Already sorted'],
        ['arr' => [1, 3, 2, 4, 5], 'desc' => 'Nearly sorted'],
    ];

    echo "\n📋 Basic Sorting Tests:\n";
    echo str_repeat("-", 60) . "\n";

    foreach ($testCases as $tc) {
        $standardResult = InsertionSort::sort($tc['arr']);
        $binaryResult = InsertionSort::binaryInsertionSort($tc['arr']);
        $shellResult = InsertionSort::shellSort($tc['arr']);
        $recursiveResult = InsertionSort::sortRecursive($tc['arr']);

        echo "\nTest: {$tc['desc']}\n";
        echo "Original: [" . implode(", ", $tc['arr']) . "]\n";
        echo "Sorted:   [" . implode(", ", $standardResult) . "]\n";

        $allCorrect = InsertionSort::isSorted($standardResult) &&
                     InsertionSort::isSorted($binaryResult) &&
                     InsertionSort::isSorted($shellResult) &&
                     InsertionSort::isSorted($recursiveResult);
        $allEqual = ($standardResult == $binaryResult) &&
                   ($standardResult == $shellResult) &&
                   ($standardResult == $recursiveResult);

        $status = ($allCorrect && $allEqual) ? "✓" : "✗";
        echo "All implementations match: $status\n";
    }

    // Visualization demo
    echo "\n\n🎬 Step-by-Step Visualization:\n";
    echo str_repeat("-", 60) . "\n";

    $demoArr = [5, 2, 8, 6, 1];
    $steps = InsertionSort::visualizeInsertionSort($demoArr);
    foreach ($steps as $step) {
        echo "$step\n";
    }

    // Stability demonstration
    echo "\n\n🔒 Stability Demonstration:\n";
    echo str_repeat("-", 60) . "\n";
    InsertionSort::demonstrateStability();

    // Performance analysis
    echo "\n\n📊 Operation Counting:\n";
    echo str_repeat("-", 60) . "\n";

    $statTestCases = [
        ['arr' => [5, 2, 8, 6, 1], 'desc' => 'Random'],
        ['arr' => [1, 2, 3, 4, 5], 'desc' => 'Already sorted'],
        ['arr' => [5, 4, 3, 2, 1], 'desc' => 'Reverse sorted'],
    ];

    foreach ($statTestCases as $tc) {
        $stats = new SortStatistics();
        InsertionSort::sortWithStats($tc['arr'], $stats);

        $n = count($tc['arr']);
        echo "\n{$tc['desc']}: [" . implode(", ", $tc['arr']) . "]\n";
        echo "Array size (n): $n\n";
        echo "Comparisons: {$stats->comparisons}\n";
        echo "Swaps: {$stats->swaps}\n";
        echo "Best case comparisons: " . ($n - 1) . "\n";
        echo "Worst case comparisons: " . ($n * ($n - 1) / 2) . "\n";
    }
}

// MARK: - Performance Benchmark

function performanceBenchmark(): void
{
    echo "\n\n⚡ Performance Benchmark\n";
    echo str_repeat("=", 80) . "\n";
    echo "\nInsertion sort is preferred for:\n";
    echo "  • Small arrays (typically n < 10-20)\n";
    echo "  • Nearly sorted arrays\n";
    echo "  • As part of hybrid sorting algorithms\n";
    echo "\n";

    $sizes = [5, 10, 20, 50, 100, 500, 1000];

    $patterns = [
        'Random' => function($n) {
            $arr = [];
            for ($i = 0; $i < $n; $i++) {
                $arr[] = rand(1, 1000);
            }
            return $arr;
        },
        'Nearly Sorted' => function($n) {
            $arr = range(0, $n - 1);
            for ($i = 0; $i < min(5, intdiv($n, 10)); $i++) {
                $idx1 = rand(0, $n - 1);
                $idx2 = rand(0, $n - 1);
                $temp = $arr[$idx1];
                $arr[$idx1] = $arr[$idx2];
                $arr[$idx2] = $temp;
            }
            return $arr;
        },
        'Reversed' => function($n) {
            return range($n - 1, 0);
        },
    ];

    foreach ($patterns as $patternName => $patternGen) {
        echo "\n$patternName Data:\n";
        printf("%-8s%15s%15s%15s%15s\n", "Size", "Insertion", "Binary", "Shell", "sort()");
        echo str_repeat("-", 68) . "\n";

        foreach ($sizes as $size) {
            $testData = $patternGen($size);
            printf("%-8d", $size);

            // Insertion Sort
            $start = microtime(true);
            InsertionSort::sort($testData);
            $elapsed = (microtime(true) - $start) * 1000;
            printf("%14.3fms", $elapsed);

            // Binary Insertion Sort
            $start = microtime(true);
            InsertionSort::binaryInsertionSort($testData);
            $elapsed = (microtime(true) - $start) * 1000;
            printf("%14.3fms", $elapsed);

            // Shell Sort
            $start = microtime(true);
            InsertionSort::shellSort($testData);
            $elapsed = (microtime(true) - $start) * 1000;
            printf("%14.3fms", $elapsed);

            // PHP sort()
            $testData2 = $testData;
            $start = microtime(true);
            sort($testData2);
            $elapsed = (microtime(true) - $start) * 1000;
            printf("%14.3fms\n", $elapsed);
        }
    }
}

// MARK: - Main

if (php_sapi_name() === 'cli') {
    demonstrateInsertionSort();
    performanceBenchmark();

    echo "\n✨ Insertion Sort demonstration complete!\n";
}

File: sorting/insertionsort.rb
# Insertion Sort Algorithm Implementation in Ruby
#
# Time Complexity:
# - Best Case: O(n) - when array is already sorted
# - Average Case: O(n²)
# - Worst Case: O(n²) - when array is reverse sorted
# Space Complexity: O(1) for in-place, O(n) for functional approach
#
# Insertion Sort builds the final sorted array one item at a time.
#
# Ruby features:
# - Idiomatic Ruby method naming (snake_case)
# - Module organization
# - Blocks and lambdas for custom comparators
# - Mixins for extension methods
# - Symbols and hash-based options

module AlgorithmsMultiverse
  module Sorting
    # MARK: - Basic Insertion Sort

    # Performs standard insertion sort.
    #
    # Time Complexity: O(n²) average and worst case, O(n) best case
    # Space Complexity: O(n) for the new array
    #
    # Example visualization:
    #   Initial: [5, 2, 8, 6, 1]
    #   Step 1:  [2, 5, 8, 6, 1]  # Insert 2
    #   Step 2:  [2, 5, 8, 6, 1]  # 8 already in place
    #   Step 3:  [2, 5, 6, 8, 1]  # Insert 6
    #   Step 4:  [1, 2, 5, 6, 8]  # Insert 1
    #
    # @param arr [Array] The array to sort
    # @param comparator [Proc, nil] Optional comparison block/proc
    # @return [Array] A new sorted array
    def self.insertion_sort(arr, &comparator)
      return arr.dup if arr.length <= 1

      result = arr.dup
      insertion_sort_in_place!(result, &comparator)
      result
    end

    # Sorts the array in-place using insertion sort.
    #
    # Time Complexity: O(n²) average and worst case, O(n) best case
    # Space Complexity: O(1)
    #
    # @param arr [Array] The array to sort (modified in-place)
    # @param comparator [Proc, nil] Optional comparison block/proc
    # @return [Array] The sorted array
    def self.insertion_sort_in_place!(arr, &comparator)
      comparator ||= ->(a, b) { a <=> b }

      (1...arr.length).each do |i|
        key = arr[i]
        j = i - 1

        # Move elements greater than key one position ahead
        while j >= 0 && comparator.call(arr[j], key) > 0
          arr[j + 1] = arr[j]
          j -= 1
        end

        arr[j + 1] = key
      end

      arr
    end

    # MARK: - Recursive Insertion Sort

    # Performs recursive insertion sort.
    #
    # Time Complexity: O(n²)
    # Space Complexity: O(n) for recursion stack
    #
    # @param arr [Array] The array to sort
    # @return [Array] A new sorted array
    def self.insertion_sort_recursive(arr)
      return arr.dup if arr.length <= 1

      result = arr.dup
      insertion_sort_recursive_helper!(result, result.length)
      result
    end

    # Helper function for recursive insertion sort.
    #
    # @param arr [Array] The array to sort
    # @param n [Integer] The number of elements to sort
    def self.insertion_sort_recursive_helper!(arr, n)
      # Base case
      return if n <= 1

      # Sort first n-1 elements
      insertion_sort_recursive_helper!(arr, n - 1)

      # Insert last element at its correct position
      key = arr[n - 1]
      j = n - 2

      while j >= 0 && arr[j] > key
        arr[j + 1] = arr[j]
        j -= 1
      end

      arr[j + 1] = key
    end

    # MARK: - Binary Insertion Sort

    # Uses binary search to find insertion position.
    #
    # Time Complexity: O(n²) for moves, O(n log n) for comparisons
    # Space Complexity: O(n)
    #
    # @param arr [Array] The array to sort
    # @return [Array] A new sorted array
    def self.binary_insertion_sort(arr)
      return arr.dup if arr.length <= 1

      result = arr.dup

      (1...result.length).each do |i|
        key = result[i]

        # Find position using binary search
        pos = binary_search_position(result, 0, i - 1, key)

        # Shift elements to make space
        (i - 1).downto(pos) do |j|
          result[j + 1] = result[j]
        end

        result[pos] = key
      end

      result
    end

    # Binary search to find the correct insertion position.
    #
    # @param arr [Array] The array
    # @param left [Integer] Left boundary
    # @param right [Integer] Right boundary
    # @param key The key to insert
    # @return [Integer] The position to insert
    def self.binary_search_position(arr, left, right, key)
      return (key > arr[left] ? left + 1 : left) if right <= left

      mid = (left + right) / 2

      return mid + 1 if key == arr[mid]

      if key > arr[mid]
        binary_search_position(arr, mid + 1, right, key)
      else
        binary_search_position(arr, left, mid - 1, key)
      end
    end

    # MARK: - Shell Sort

    # Performs shell sort (generalization of insertion sort).
    #
    # Time Complexity: Depends on gap sequence (O(n log²n) for good sequences)
    # Space Complexity: O(n)
    #
    # @param arr [Array] The array to sort
    # @return [Array] A new sorted array
    def self.shell_sort(arr)
      return arr.dup if arr.length <= 1

      result = arr.dup
      n = result.length

      # Start with a large gap, then reduce (Knuth's sequence)
      gap = 1
      gap = 3 * gap + 1 while gap < n / 3

      # Perform gapped insertion sort
      while gap > 0
        (gap...n).each do |i|
          key = result[i]
          j = i

          # Insertion sort with gap
          while j >= gap && result[j - gap] > key
            result[j] = result[j - gap]
            j -= gap
          end

          result[j] = key
        end

        gap /= 3
      end

      result
    end

    # MARK: - Sort Statistics

    # Tracks sorting operations
    class SortStatistics
      attr_accessor :comparisons, :swaps

      def initialize
        @comparisons = 0
        @swaps = 0
      end

      def reset
        @comparisons = 0
        @swaps = 0
      end

      def to_s
        "Comparisons: #{@comparisons}, Swaps: #{@swaps}"
      end
    end

    # Insertion sort with statistics tracking
    #
    # @param arr [Array] The array to sort
    # @param stats [SortStatistics] Statistics object to track operations
    # @return [Array] A new sorted array
    def self.insertion_sort_with_stats(arr, stats)
      stats.reset
      return arr.dup if arr.length <= 1

      result = arr.dup

      (1...result.length).each do |i|
        key = result[i]
        j = i - 1

        while j >= 0
          stats.comparisons += 1
          if result[j] > key
            result[j + 1] = result[j]
            stats.swaps += 1
            j -= 1
          else
            break
          end
        end

        result[j + 1] = key
      end

      result
    end

    # MARK: - Visualization

    # Creates a step-by-step visualization of insertion sort
    #
    # @param arr [Array<Integer>] The array to visualize
    # @return [Array<String>] Array of strings showing each step
    def self.visualize_insertion_sort(arr)
      steps = []
      result = arr.dup

      steps << "Initial: #{result}"

      (1...result.length).each do |i|
        key = result[i]
        j = i - 1

        steps << "\nStep #{i}: Inserting #{key}"
        steps << "  Before: #{result}"

        while j >= 0 && result[j] > key
          result[j + 1] = result[j]
          j -= 1
        end

        result[j + 1] = key
        steps << "  After:  #{result}"
      end

      steps << "\nFinal: #{result}"
      steps
    end

    # MARK: - Stability Demonstration

    # Demonstrates that insertion sort is stable
    def self.demonstrate_stability
      Pair = Struct.new(:value, :original_index)

      data = [
        Pair.new(3, 0),
        Pair.new(1, 1),
        Pair.new(3, 2),
        Pair.new(2, 3),
        Pair.new(3, 4)
      ]

      # Sort by value only
      sorted = insertion_sort(data) { |a, b| a.value <=> b.value }

      puts "Stability Demonstration:"
      print "Original: "
      data.each { |p| print "(#{p.value},#{p.original_index}) " }
      puts

      print "Sorted:   "
      sorted.each { |p| print "(#{p.value},#{p.original_index}) " }
      puts

      # Check stability - all 3's should maintain original order
      three_indices = sorted.select { |p| p.value == 3 }.map(&:original_index)
      is_stable = three_indices == [0, 2, 4]
      puts "Stable: #{is_stable} (indices of 3's: #{three_indices})"
    end

    # MARK: - Helper Functions

    # Check if an array is sorted in ascending order
    #
    # @param arr [Array] The array to check
    # @return [Boolean] True if sorted, false otherwise
    def self.sorted?(arr)
      (0...arr.length - 1).all? { |i| arr[i] <= arr[i + 1] }
    end
  end
end

# MARK: - Array Extension

class Array
  # Returns a new array sorted using insertion sort
  #
  # @param comparator [Proc, nil] Optional comparison block
  # @return [Array] A new sorted array
  def insertion_sorted(&comparator)
    AlgorithmsMultiverse::Sorting.insertion_sort(self, &comparator)
  end

  # Sorts the array in-place using insertion sort
  #
  # @param comparator [Proc, nil] Optional comparison block
  # @return [Array] Self
  def insertion_sort!(&comparator)
    AlgorithmsMultiverse::Sorting.insertion_sort_in_place!(self, &comparator)
  end

  # Check if array is sorted
  #
  # @return [Boolean] True if sorted
  def sorted?
    AlgorithmsMultiverse::Sorting.sorted?(self)
  end
end

# MARK: - Demonstration and Testing

def demonstrate_insertion_sort
  include AlgorithmsMultiverse::Sorting

  puts "📝 Insertion Sort Implementation in Ruby"
  puts "=" * 60

  # Test data
  test_cases = [
    { arr: [64, 34, 25, 12, 22, 11, 90], desc: "Random array" },
    { arr: [5, 2, 8, 6, 1, 9, 4], desc: "Small random array" },
    { arr: [1], desc: "Single element" },
    { arr: [], desc: "Empty array" },
    { arr: [3, 3, 3, 3, 3], desc: "All duplicates" },
    { arr: [9, 8, 7, 6, 5, 4, 3, 2, 1], desc: "Reverse sorted" },
    { arr: [1, 2, 3, 4, 5], desc: "Already sorted" },
    { arr: [1, 3, 2, 4, 5], desc: "Nearly sorted" }
  ]

  puts "\n📋 Basic Sorting Tests:"
  puts "-" * 60

  test_cases.each do |tc|
    standard_result = AlgorithmsMultiverse::Sorting.insertion_sort(tc[:arr])
    binary_result = AlgorithmsMultiverse::Sorting.binary_insertion_sort(tc[:arr])
    shell_result = AlgorithmsMultiverse::Sorting.shell_sort(tc[:arr])
    recursive_result = AlgorithmsMultiverse::Sorting.insertion_sort_recursive(tc[:arr])

    puts "\nTest: #{tc[:desc]}"
    puts "Original: #{tc[:arr]}"
    puts "Sorted:   #{standard_result}"

    all_correct = standard_result.sorted? && binary_result.sorted? &&
                 shell_result.sorted? && recursive_result.sorted?
    all_equal = (standard_result == binary_result) &&
               (standard_result == shell_result) &&
               (standard_result == recursive_result)

    status = (all_correct && all_equal) ? "✓" : "✗"
    puts "All implementations match: #{status}"
  end

  # Visualization demo
  puts "\n\n🎬 Step-by-Step Visualization:"
  puts "-" * 60

  demo_arr = [5, 2, 8, 6, 1]
  steps = AlgorithmsMultiverse::Sorting.visualize_insertion_sort(demo_arr)
  steps.each { |step| puts step }

  # Stability demonstration
  puts "\n\n🔒 Stability Demonstration:"
  puts "-" * 60
  AlgorithmsMultiverse::Sorting.demonstrate_stability

  # Performance analysis
  puts "\n\n📊 Operation Counting:"
  puts "-" * 60

  stat_test_cases = [
    { arr: [5, 2, 8, 6, 1], desc: "Random" },
    { arr: [1, 2, 3, 4, 5], desc: "Already sorted" },
    { arr: [5, 4, 3, 2, 1], desc: "Reverse sorted" }
  ]

  stat_test_cases.each do |tc|
    stats = AlgorithmsMultiverse::Sorting::SortStatistics.new
    AlgorithmsMultiverse::Sorting.insertion_sort_with_stats(tc[:arr], stats)

    n = tc[:arr].length
    puts "\n#{tc[:desc]}: #{tc[:arr]}"
    puts "Array size (n): #{n}"
    puts "Comparisons: #{stats.comparisons}"
    puts "Swaps: #{stats.swaps}"
    puts "Best case comparisons: #{n - 1}"
    puts "Worst case comparisons: #{n * (n - 1) / 2}"
  end
end

# MARK: - Performance Benchmark

def performance_benchmark
  require 'benchmark'
  include AlgorithmsMultiverse::Sorting

  puts "\n\n⚡ Performance Benchmark"
  puts "=" * 80
  puts "\nInsertion sort is preferred for:"
  puts "  • Small arrays (typically n < 10-20)"
  puts "  • Nearly sorted arrays"
  puts "  • As part of hybrid sorting algorithms"
  puts

  sizes = [5, 10, 20, 50, 100, 500, 1000]

  patterns = {
    'Random' => ->(n) { Array.new(n) { rand(1..1000) } },
    'Nearly Sorted' => lambda { |n|
      arr = (0...n).to_a
      [5, n / 10].min.times do
        idx1, idx2 = rand(n), rand(n)
        arr[idx1], arr[idx2] = arr[idx2], arr[idx1]
      end
      arr
    },
    'Reversed' => ->(n) { (0...n).to_a.reverse }
  }

  patterns.each do |pattern_name, pattern_gen|
    puts "\n#{pattern_name} Data:"
    puts format("%-8s%15s%15s%15s%15s", "Size", "Insertion", "Binary", "Shell", "sort()")
    puts "-" * 68

    sizes.each do |size|
      test_data = pattern_gen.call(size)
      print format("%-8d", size)

      # Insertion Sort
      elapsed = Benchmark.measure {
        AlgorithmsMultiverse::Sorting.insertion_sort(test_data)
      }.real * 1000
      print format("%14.3fms", elapsed)

      # Binary Insertion Sort
      elapsed = Benchmark.measure {
        AlgorithmsMultiverse::Sorting.binary_insertion_sort(test_data)
      }.real * 1000
      print format("%14.3fms", elapsed)

      # Shell Sort
      elapsed = Benchmark.measure {
        AlgorithmsMultiverse::Sorting.shell_sort(test_data)
      }.real * 1000
      print format("%14.3fms", elapsed)

      # Ruby sort()
      elapsed = Benchmark.measure {
        test_data.sort
      }.real * 1000
      puts format("%14.3fms", elapsed)
    end
  end
end

# MARK: - Main

if __FILE__ == $PROGRAM_NAME
  demonstrate_insertion_sort
  performance_benchmark

  puts "\n✨ Insertion Sort demonstration complete!"
end
